ui/src/app/applications/components/application-details/application-details.tsx

```diff
@@ -580,7 +590,18 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
                                                 {title: 'Applications', path: '/applications'},
                                                 {title: <ApplicationsDetailsAppDropdown appName={this.props.match.params.name} />}
                                             ],
-                                            actionMenu: {items: this.getApplicationActionMenu(application, true)},
+                                            actionMenu: {
+                                                items: [...this.getApplicationActionMenu(application, true)]
+                                            },
+                                            actionMenuExtensions: TopBarExtensions.map(ext => (
```

Comment:
this doesn't seem to require a new prop -- you can pass an additional `item` in the list above and add a callback (`item.action`) that makes this extension view visible

See also https://github.com/argoproj/argo-ui/pull/576#discussion_r1725980287

Reply:
I would like to make the extension  an opt-in  button. In the UI, additional items always show up as buttons, regardless of whether the extension is needed for that app. If I make change in argo,  I can control the content inside button but not the button, it will always render on the UI.
 For my use case, the button should be displayed when the application has specific  label (prd) and hidden when non-PRD. I don't want to add any logic to determine when to render the extension. I tried the option you mentioned but it doesn't fit the requirement. Add the screenshot for the changes.

Reply:
> In the UI, additional items always show up as buttons, regardless of whether the extension is needed for that app. 

The convention for something optional would be to only add it to the items list when it is needed. Adding something that doesn't render would be an anti-pattern

> For my use case, the button should be displayed when the application has specific label (prd) and hidden when non-PRD. I don't want to add any logic to determine when to render the extension.


You have to have that logic somewhere, this just sounds like changing where that logic goes

Reply:
From you [comment](https://github.com/argoproj/argo-cd/pull/19597#issuecomment-2303378924), not sure how I can make extension customizable without adding logic in argocd. I don't want define any template for extension. If Can you provide some  directions ?

Reply:
You can have the extension expose an `isVisible` or `shouldDisplay` function (the latter is a little more accurately named). For instance, from the extension registration you have in the docs now:

```diff
- registerTopBarExtension(component: StatusPanelExtensionComponent, title: string, id: string, flyout?: ExtensionComponent, isMiddle: boolean, isNarrow: boolean)
+ registerTopBarExtension(component: StatusPanelExtensionComponent, title: string, id: string, flyout?: ExtensionComponent, isMiddle: boolean, isNarrow: boolean, shouldDisplay?: () => boolean)
```

You can add parameters to the `shouldDisplay` function if that would help as well

Reply:
thank you for the input, changes make sense to me. I make the change and test.

Reply:
@agilgur5  As you recommended, I don't rely on argo-ui change now, and only argocd has the changes. Can you review the PR?

Reply:
This approach looks a lot better to me 🙂 See my review [below](https://github.com/argoproj/argo-cd/pull/19620#pullrequestreview-2259307947)